x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/dataframe/process/DataFrameRowsJoiner.java

```diff
@@ -0,0 +1,124 @@
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+
+public class DataFrameRowsJoiner {
+
+    private static final Logger LOGGER = LogManager.getLogger(DataFrameRowsJoiner.class);
+
+    private final String analyticsId;
+    private final Client client;
+    private final DataFrameDataExtractor dataExtractor;
+    private List<DataFrameDataExtractor.Row> currentDataFrameRows;
+    private List<RowResults> currentResults;
+    private boolean failed;
+
+    public DataFrameRowsJoiner(String analyticsId, Client client, DataFrameDataExtractor dataExtractor) {
+        this.analyticsId = Objects.requireNonNull(analyticsId);
+        this.client = Objects.requireNonNull(client);
+        this.dataExtractor = Objects.requireNonNull(dataExtractor);
+    }
+
+    public void processRowResults(RowResults rowResults) {
+        if (failed) {
+            // If we are in failed state we drop the results but we let the processor
+            // parse the output
+            return;
+        }
+
+        try {
+            addResultAndJoinIfEndOfBatch(rowResults);
+        } catch (Exception e) {
+            LOGGER.error(new ParameterizedMessage("[{}] Failed to join results", analyticsId), e);
+            failed = true;
+        }
+    }
+
+    private void addResultAndJoinIfEndOfBatch(RowResults rowResults) {
+        if (currentDataFrameRows == null) {
+            Optional<List<DataFrameDataExtractor.Row>> nextBatch = getNextBatch();
+            if (nextBatch.isPresent() == false) {
+                return;
+            }
+            currentDataFrameRows = nextBatch.get();
+            currentResults = new ArrayList<>(currentDataFrameRows.size());
+        }
+        currentResults.add(rowResults);
+        if (currentResults.size() == currentDataFrameRows.size()) {
+            joinCurrentResults();
+            currentDataFrameRows = null;
+        }
+    }
+
+    private Optional<List<DataFrameDataExtractor.Row>> getNextBatch() {
+        try {
+            return dataExtractor.next();
+        } catch (IOException e) {
+            // TODO Implement recovery strategy or better error reporting
+            LOGGER.error("Error reading next batch of data frame rows", e);
+            return Optional.empty();
+        }
+    }
+
+    private void joinCurrentResults() {
+        BulkRequest bulkRequest = new BulkRequest();
+        for (int i = 0; i < currentDataFrameRows.size(); i++) {
+            DataFrameDataExtractor.Row row = currentDataFrameRows.get(i);
+            if (row.shouldSkip()) {
+                continue;
+            }
+            RowResults result = currentResults.get(i);
+            checkChecksumsMatch(row, result);
+
+            SearchHit hit = row.getHit();
+            Map<String, Object> source = new LinkedHashMap(hit.getSourceAsMap());
+            source.putAll(result.getResults());
+            IndexRequest indexRequest = new IndexRequest(hit.getIndex(), hit.getType(), hit.getId());
```

Comment:
This method is deprecated, so it would be best not to use it in new code:

```
@deprecated Types are in the process of being removed. Use {@link #IndexRequest(String)} with {@link #id(String)} instead.
```

Reply:
Done